File: core/src/de/project/ice/ecs/components/MoveComponent.java
package de.project.ice.ecs.components;

import com.badlogic.gdx.math.Vector2;

import java.util.LinkedList;
import java.util.List;

/**
 * Component for movable objects and characters
 */
public class MoveComponent implements IceComponent<MoveComponent>
{
    public boolean isMoving = false;
    public List<Vector2> targetPositions = new LinkedList<Vector2>();
    public float speed = 2.0f;

    @Override
    public void reset()
    {
        isMoving = false;
        speed = 2.0f;
        targetPositions.clear();
    }

    @Override
    public void copyTo(MoveComponent copy)
    {
        copy.isMoving = isMoving;
        copy.speed = speed;
        for (Vector2 pos : targetPositions)
        {
            copy.targetPositions.add(pos.cpy());
        }
    }
}

File: core/src/de/project/ice/ecs/components/MoveComponent.kt
package de.project.ice.ecs.components


import com.badlogic.gdx.math.Vector2
import de.project.ice.annotations.Property
import java.util.*

/**
 * Component for movable objects and characters
 */
class MoveComponent : CopyableIceComponent {
    @Property("Whether the entity is currently moving or not", true)
    var isMoving = false
    @Property("The position to where this entity is walking")
    val targetPositions = ArrayList<Vector2>()
    @Property("The moving speed of this entity")
    var speed = 2.0f
    var callback: (()->Unit)? = null;

    override fun reset() {
        isMoving = false
        speed = 2.0f
        targetPositions.clear()
        callback = null
    }

    override fun copyTo(copy: CopyableIceComponent) {
        if (copy is MoveComponent) {
            copy.isMoving = isMoving
            copy.speed = speed
            copy.callback = callback
            for (pos in targetPositions) {
                copy.targetPositions.add(pos.cpy())
            }
        }
    }
}
